compiler/rustc_codegen_ssa/src/mir/rvalue.rs

```diff
@@ -819,8 +819,14 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                     .builtin_deref(true)
                     .unwrap_or_else(|| bug!("deref of non-pointer {:?}", input_ty))
                     .ty;
-                let llty = bx.cx().backend_type(bx.cx().layout_of(pointee_type));
-                bx.inbounds_gep(llty, lhs, &[rhs])
+                let pointee_layout = bx.cx().layout_of(pointee_type);
+                if pointee_layout.is_zst() {
+                    // ZSTs are weird, since it's legal to `.add(usize::MAX)` on them.
```

Comment:
Was this possible to trigger before? What led to this change?

Reply:
Oh, this is left over from a previous change where the other branch was unsound for ZSTs, but I removed that.

So I could revert this too, but since the old code was also calculating the layout anyway, I figured it was reasonable to do the [quick](https://github.com/rust-lang/rust/blob/e3ccd4b9a51ded19bf1c3f6c647c96b6dd52eced/compiler/rustc_abi/src/lib.rs#L1565-L1571) ZST check ourselves rather than giving LLVM the useless GEP.

I'm happy to go either way; let me know what you'd prefer.

Reply:
We can keep it.